### PR TITLE
Add bank corner plot

### PR DESCRIPTION
**Pull request by GarethCabournDavies**
Add a corner plot for template banks, using a combination of `bank_conversions.py` from `tmpltbank` and the `create_multidim_plot` function used in inference results plotting

Some examples in https://ldas-jobs.ligo.caltech.edu/~gareth.cabourndavies/testoutput/add_bank_corner_plot/testing/, with the bash script used to create them (and some specific failure cases) and the output of the script in the .err file

### CODE REVIEW

**Comment by GarethCabournDavies** [comment]
Poke @titodalcanton, though if youre busy / dont feel able to comment on this code, then someone else may be able to review

File: bin/plotting/pycbc_plot_bank_corner [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+from pycbc.tmpltbank import bank_conversions as bconv
+from pycbc.pnutils import get_imr_duration
+from pycbc.results.scatter_histograms import create_multidim_plot
+from pycbc.results import metadata
+
+conversion_options = bconv.conversion_options
+
+parser = argparse.ArgumentParser(usage='pycbc_plot_bank_corner [--options]',
+                    description='Plots various parameters against one '
+                          'another for pycbc banks in an hdf file.')
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_bank_corner:38]: There is a "feed two birds with one scone" case here: add a docstring to the entire file with this description, then just use `description=__doc__` when you instantiate the ArgumentParser.

File: bin/plotting/pycbc_plot_bank_corner [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+         "the smallest parameter value in the posterior "
+         "will be used.")
+parser.add_argument('--maxs',
+    nargs='+',
+    metavar='PARAM:VAL',
+    default=[],
+    help="Same as mins, but for the maximum values to "
+         "plot.")
+parser.add_argument("--color-parameter",
+    nargs=1,
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_bank_corner:97]: Why is this needed if you only take 1 parameter?

File: bin/plotting/pycbc_plot_bank_corner [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+bank = {}
+with h5py.File(args.bank_file, 'r') as bankf:
+    for p in bankf.keys():
+        if not isinstance(bankf[p], h5py.Dataset): continue
+        # Ignore things which aren't numbers, which cannot
+        # be histogrammed - e.g. approximant
+        if not np.issubdtype(bankf[p].dtype, np.number): continue
+        bank[p] = bankf[p][:]
+
+banklen = list(bank.values())[0].size
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_bank_corner:129]: This may require more RAM than is reasonable for such an operation. Can you instead set `banklen` in the `for` loop above like `banklen = len(bankf[p])`?

File: bin/plotting/pycbc_plot_bank_corner [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+logging.info("Got %d templates from the bank", banklen)
+
+# If no parameters are given - just plot whatever is in the bank
+if not args.parameters:
+    # Plot anything which is not singular
+    args.parameters = [k for k in bank.keys()
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_bank_corner:135]:
```suggestion
    args.parameters = [k for k in bank
```

File: bin/plotting/pycbc_plot_bank_corner [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+                       if len(np.unique(bank[k])) > 1]
+    args.parameters_labels = {k: k for k in list(bank.keys())}
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_bank_corner:137]: Can this be simplified like so:
```suggestion
    args.parameters_labels = {k: k for k in bank}
```

Comment (titodalcanton) [reply bin/plotting/pycbc_plot_bank_corner:137]: Also, one may perhaps want to not give args.parameters but still give args.parameters_labels?

File: bin/plotting/pycbc_plot_bank_corner [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    if len(sharex_axes) > 1:
+        for s0, s1 in zip(sharex_axes[:-1], sharex_axes[1:]):
+            s0.sharex(s1)
+
+logging.info("Plot generated")
+fig.set_dpi(args.dpi)
+
+# save
+caption = "Template bank as a corner plot with "
+caption += "scatter points for each waveform."
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_bank_corner:276]:
```suggestion
caption = ("Template bank as a corner plot with "
           "scatter points for each waveform.")
```

File: bin/plotting/pycbc_plot_bank_corner [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+
+    logging.info("Converting %s", p)
+    bank[p] = bconv.get_bank_property(p, bank, np.arange(banklen))
+
+# Durations can be negative where get_imr_duration is not perfect - 
+# set a minimum value of 0.15s
+if 'duration' in bank:
+    bank['duration'] = np.maximum(bank['duration'], 0.15)
+if 'template_duration' in bank:
+    bank['template_duration'] = np.maximum(bank['template_duration'], 0.15)
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_bank_corner:191]: So is there no way of plotting durations shorter than 150 ms, even if computed correctly? I would say that a plotting code should just plot whatever comes out of the duration calculation…

File: bin/plotting/pycbc_plot_bank_corner [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+if args.mpl_style == 'dark_background':
+    hist_color = 'white'
+else:
+    hist_color = 'black'
+
+# All parameters should be in the bank now, check they are the right size:
+assert all([len(bank[p]) == banklen for p in bank])
+
+# Change the bank to the format required for the plot:
+bank_fa = FieldArray.from_arrays(bank.values(), names=list(bank.keys()))
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_bank_corner:202]: It is guarantees that `bank.values()` and `bank.keys()` will always return keys and values in a consistent order?

File: bin/plotting/pycbc_plot_bank_corner [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+logging.info("Generating corner plot")
+
+fig, axis_dict = create_multidim_plot(
+    args.parameters,
+    bank_fa,
+    labels=args.parameters_labels,
+    plot_marginal=args.plot_histogram,
+    plot_scatter=True,
+    plot_contours=False,
+    scatter_cmap="rainbow",
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_bank_corner:228]: I would avoid the rainbow colormap because of its [weird lightness behavior](https://matplotlib.org/stable/_images/sphx_glr_colormaps_014.png). Use one of the [perceptually uniform](https://matplotlib.org/stable/_images/sphx_glr_colormaps_008.png) colormaps instead.

File: bin/plotting/pycbc_plot_bank_corner [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+    show_colorbar=cpar is not False,
+    cbar_label=cpar_label,
+    vmin=mins[cpar] if cpar else 0,
+    vmax=maxs[cpar] if cpar else 0,
+    hist_color=hist_color,
+    mins=mins,
+    maxs=maxs
+)
+
+fig.suptitle(f"Bank length: {banklen}")
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_bank_corner:242]: Suggestion for an alternative title:
```suggestion
fig.suptitle(f"{os.path.basename(args.bank_file)} - {banklen} templates")
```

File: pycbc/tmpltbank/bank_conversions.py [hunk 11]
```diff
@@ -88,20 +83,56 @@ def get_bank_property(parameter, bank, template_ids,
...
         values = bank[parameter][:][template_ids]
-
-    # These things may be in the bank, but if not, we need to calculate
-    elif parameter in ['template_duration', 'duration']:
-        if 'template_duration' in bank:
+    # Duration may be in the bank, but if not, we need to calculate
+    elif parameter.endswith('duration'):
+        duration_approximant = bank['template_duration'][:][template_ids] \
+            if 'approximant' in bank \
+            else np.repeat("SPAtmplt", len(template_ids))
```
Comment (titodalcanton) [inline pycbc/tmpltbank/bank_conversions.py:90]: This is effectively changing the default approximant for duration calculation from SEOBNRv4 to SPAtmplt. Not sure this is necessarily a good idea, and if it is, a plotting PR is probably not the place to do it. Can you keep the `duration_approximant` as a kwarg with the current default?

Comment (titodalcanton) [reply pycbc/tmpltbank/bank_conversions.py:90]: Also, I guess line 88 meant to use `approximant` as index, not `template_duration`?

File: bin/plotting/pycbc_plot_bank_corner (hunk 7 above)
Comment (GarethCabournDavies) [reply bin/plotting/pycbc_plot_bank_corner:191]: This was just a way of getting around the negative duration issue, and was valid for the purpose I used it for, then forgot to fix it. I'm not sure what is best to say to do for this though

Comment (titodalcanton) [reply bin/plotting/pycbc_plot_bank_corner:191]: Treat the duration as any other value. If the user tries to use a log duration scale, and some durations are negative, the code will produce an error. This seems a perfectly reasonable behavior to me.

File: bin/plotting/pycbc_plot_bank_corner (hunk 8 above)
Comment (GarethCabournDavies) [reply bin/plotting/pycbc_plot_bank_corner:202]: > Changed in version 3.7: Dictionary order is guaranteed to be insertion order

From the [documentation](https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects)

File: bin/plotting/pycbc_plot_bank_corner (hunk 9 above)
Comment (GarethCabournDavies) [reply bin/plotting/pycbc_plot_bank_corner:228]: Updated to use plasma - many others have points with brightness that goes too close to white (in my opinion)

**Comment by GarethCabournDavies** [comment]
I think this is ready for re-review / merge now

**Comment by titodalcanton** [comment]
@GarethCabournDavies could you please rerun your examples with the latest version?

**Comment by GarethCabournDavies** [comment]
> @GarethCabournDavies could you please rerun your examples with the latest version?

See https://ldas-jobs.ligo.caltech.edu/~gareth.cabourndavies/testoutput/add_bank_corner_plot/testing/2023-06-13/
